sequencer/src/block.rs

```diff
@@ -17,28 +19,131 @@ use serde::{Deserialize, Deserializer, Serialize, Serializer};
 use std::{
     env,
     fmt::{Debug, Display},
+    ops::Add,
     time::Duration,
 };
 use time::OffsetDateTime;
-use typenum::U3;
+use typenum::Unsigned;
+use typenum::U2;
 
 pub type BlockMerkleTree = LightWeightSHA3MerkleTree<Commitment<Header>>;
 pub type BlockMerkleCommitment = <BlockMerkleTree as MerkleTreeScheme>::Commitment;
 
-#[derive(Default)]
+// New Type for `U256` in order to implement `CanonicalSerialize` and
+// `CanonicalDeserialize`
+#[derive(Default, Hash, Copy, Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct FeeElement(U256);
+// New Type for `Address` in order to implement `CanonicalSerialize` and
+// `CanonicalDeserialize`
+#[derive(
+    Default, Hash, Copy, Clone, Debug, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord,
+)]
+pub struct FeeIndex(Address);
+
+impl Add for FeeElement {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Self(self.0.add(other.0))
+    }
+}
+
+impl Valid for FeeElement {
+    fn check(&self) -> Result<(), SerializationError> {
+        Ok(())
+    }
+}
+
+impl Valid for FeeIndex {
+    fn check(&self) -> Result<(), SerializationError> {
+        Ok(())
+    }
+}
+
+impl CanonicalSerialize for FeeElement {
+    fn serialize_with_mode<W: std::io::prelude::Write>(
+        &self,
+        mut writer: W,
+        _compress: Compress,
+    ) -> Result<(), SerializationError> {
+        let mut bytes = [0u8; core::mem::size_of::<U256>()];
+        self.0.to_little_endian(&mut bytes);
+        Ok(writer.write_all(&bytes)?)
+    }
+
+    fn serialized_size(&self, _compress: Compress) -> usize {
+        core::mem::size_of::<U256>()
+    }
+}
+impl CanonicalDeserialize for FeeElement {
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        _compress: Compress,
+        _validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let mut bytes = [0u8; core::mem::size_of::<U256>()];
+        // let mut ret = [0u8; 32];
+        reader.read_exact(&mut bytes)?;
+        let value = U256::from_little_endian(&bytes);
+        Ok(Self(value))
+    }
+}
+impl CanonicalSerialize for FeeIndex {
+    fn serialize_with_mode<W: std::io::prelude::Write>(
+        &self,
+        mut writer: W,
+        _chompress: Compress,
+    ) -> Result<(), SerializationError> {
+        Ok(writer.write_all(self.0.as_bytes())?)
+    }
+
+    fn serialized_size(&self, _compress: Compress) -> usize {
+        core::mem::size_of::<Address>()
+    }
+}
+impl CanonicalDeserialize for FeeIndex {
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        _compress: Compress,
+        _validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let mut bytes = [0u8; core::mem::size_of::<Address>()];
+        reader.read_exact(&mut bytes)?;
+        let value = Address::from_slice(&bytes);
+        Ok(Self(value))
+    }
+}
+impl std::convert::From<u64> for FeeIndex {
```

Comment:
Why do we want this conversion?

Reply:
It is a requirement of `UniversalMerkleTreeScheme`: https://github.com/EspressoSystems/jellyfish/blob/bf1a6dfb63e2ce9b409a682d2793b836f1659097/primitives/src/merkle_tree/universal_merkle_tree.rs#L34

Reply:
Ah right. The thing that needs to be implemented but doesn't really get used. Np